File: frontend/src/apis/work-queue.ts
// SPDX-FileCopyrightText: 2025 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

import { apiClient } from './client'

// Enums
export type QueueVisibility = 'private' | 'public' | 'group_visible' | 'invite_only'
export type QueueMessageStatus = 'unread' | 'read' | 'processing' | 'processed' | 'archived'
export type QueueMessagePriority = 'low' | 'normal' | 'high'
export type TriggerMode = 'immediate' | 'manual' | 'scheduled' | 'condition_based'
export type ConditionType = 'priority_high' | 'specific_sender'
export type ConditionAction = 'immediate' | 'skip'

// Types
export interface TeamRef {
  namespace: string
  name: string
}

export interface ProcessCondition {
  type: ConditionType
  value?: string
  action: ConditionAction
}

export interface AutoProcessConfig {
  enabled: boolean
  teamRef?: TeamRef
  triggerMode: TriggerMode
  scheduleInterval?: number
  conditions?: ProcessCondition[]
}

export interface ResultFeedbackConfig {
  replyToSender: boolean
  saveInQueue: boolean
  sendNotification: boolean
}

export interface WorkQueue {
  id: number
  name: string
  displayName: string
  description?: string
  isDefault: boolean
  visibility: QueueVisibility
  visibleToGroups?: string[]
  inviteCode?: string
  autoProcess?: AutoProcessConfig
  resultFeedback?: ResultFeedbackConfig
  messageCount: number
  unreadCount: number
  createdAt: string
  updatedAt: string
}

export interface WorkQueueListResponse {
  items: WorkQueue[]
  total: number
}

export interface WorkQueueCreateRequest {
  name: string
  displayName: string
  description?: string
  visibility?: QueueVisibility
  visibleToGroups?: string[]
  autoProcess?: AutoProcessConfig
  resultFeedback?: ResultFeedbackConfig
}

export interface WorkQueueUpdateRequest {
  displayName?: string
  description?: string
  visibility?: QueueVisibility
  visibleToGroups?: string[]
  autoProcess?: AutoProcessConfig
  resultFeedback?: ResultFeedbackConfig
}

// Queue Message Types
export interface SenderInfo {
  id: number
  userName: string
  email?: string
}

export interface MessageContentSnapshot {
  role: string
  content: string
  senderUserName?: string
  createdAt?: string
  attachments?: Array<{
    id: number
    context_type: string
    name: string
    status: string
    file_extension?: string
    file_size?: number
  }>
}

export interface QueueMessage {
  id: number
  queueId: number
  sender: SenderInfo
  sourceTaskId: number
  sourceSubtaskIds: number[]
  contentSnapshot: MessageContentSnapshot[]
  note?: string
  priority: QueueMessagePriority
  status: QueueMessageStatus
  processResult?: Record<string, unknown>
  processTaskId?: number
  createdAt: string
  updatedAt: string
  processedAt?: string
}

export interface QueueMessageListResponse {
  items: QueueMessage[]
  total: number
  unreadCount: number
}

export interface UnreadCountResponse {
  total: number
  byQueue: Record<number, number>
}

// Forward Message Types
export interface ForwardRecipient {
  type: 'user' | 'group'
  id: number
  queueId?: number
}

export interface ForwardMessageRequest {
  sourceTaskId: number
  subtaskIds?: number[]
  recipients: ForwardRecipient[]
  note?: string
  priority?: QueueMessagePriority
}

export interface ForwardMessageResponse {
  success: boolean
  forwardedCount: number
  failedRecipients?: Array<{
    type: string
    id: number
    error: string
  }>
}

// Public Queue Types
export interface PublicQueue {
  id: number
  name: string
  displayName: string
  description?: string
  isDefault: boolean
}

export interface UserPublicQueuesResponse {
  userId: number
  userName: string
  queues: PublicQueue[]
}

// Recent Contact Types
export interface RecentContact {
  id: number
  userId: number
  userName: string
  email?: string
  lastContactAt: string
  contactCount: number
}

export interface RecentContactsListResponse {
  items: RecentContact[]
  total: number
}

// API Functions

// Work Queue Management
export async function listWorkQueues(): Promise<WorkQueueListResponse> {
  return apiClient.get<WorkQueueListResponse>('/work-queues')
}

export async function getWorkQueue(queueId: number): Promise<WorkQueue> {
  return apiClient.get<WorkQueue>(`/work-queues/${queueId}`)
}

export async function createWorkQueue(data: WorkQueueCreateRequest): Promise<WorkQueue> {
  return apiClient.post<WorkQueue>('/work-queues', data)
}

export async function updateWorkQueue(
  queueId: number,
  data: WorkQueueUpdateRequest
): Promise<WorkQueue> {
  return apiClient.put<WorkQueue>(`/work-queues/${queueId}`, data)
}

export async function deleteWorkQueue(queueId: number): Promise<void> {
  return apiClient.delete(`/work-queues/${queueId}`)
}

export async function setDefaultQueue(queueId: number): Promise<WorkQueue> {
  return apiClient.post<WorkQueue>(`/work-queues/${queueId}/set-default`)
}

export async function regenerateInviteCode(queueId: number): Promise<WorkQueue> {
  return apiClient.post<WorkQueue>(`/work-queues/${queueId}/regenerate-invite`)
}

// Queue Messages
export async function listQueueMessages(
  queueId: number,
  params?: {
    status?: QueueMessageStatus
    priority?: QueueMessagePriority
    sender_user_id?: number
    skip?: number
    limit?: number
    sort_by?: string
    sort_order?: 'asc' | 'desc'
  }
): Promise<QueueMessageListResponse> {
  const searchParams = new URLSearchParams()
  if (params?.status) searchParams.append('status', params.status)
  if (params?.priority) searchParams.append('priority', params.priority)
  if (params?.sender_user_id) searchParams.append('sender_user_id', String(params.sender_user_id))
  if (params?.skip !== undefined) searchParams.append('skip', String(params.skip))
  if (params?.limit !== undefined) searchParams.append('limit', String(params.limit))
  if (params?.sort_by) searchParams.append('sort_by', params.sort_by)
  if (params?.sort_order) searchParams.append('sort_order', params.sort_order)

  const query = searchParams.toString()
  const url = `/work-queues/${queueId}/messages${query ? `?${query}` : ''}`
  return apiClient.get<QueueMessageListResponse>(url)
}

export async function getQueueMessage(messageId: number): Promise<QueueMessage> {
  return apiClient.get<QueueMessage>(`/queue-messages/${messageId}`)
}

export async function updateMessageStatus(
  messageId: number,
  status: QueueMessageStatus
): Promise<QueueMessage> {
  return apiClient.patch<QueueMessage>(`/queue-messages/${messageId}/status`, { status })
}

export async function updateMessagePriority(
  messageId: number,
  priority: QueueMessagePriority
): Promise<QueueMessage> {
  return apiClient.patch<QueueMessage>(`/queue-messages/${messageId}/priority`, { priority })
}

export async function deleteQueueMessage(messageId: number): Promise<void> {
  return apiClient.delete(`/queue-messages/${messageId}`)
}

export async function getUnreadMessageCount(): Promise<UnreadCountResponse> {
  return apiClient.get<UnreadCountResponse>('/work-queues/messages/unread-count')
}

// Message Forwarding
export async function forwardMessages(data: ForwardMessageRequest): Promise<ForwardMessageResponse> {
  return apiClient.post<ForwardMessageResponse>('/messages/forward', data)
}

// User Public Queues
export async function getUserPublicQueues(userId: number): Promise<UserPublicQueuesResponse> {
  return apiClient.get<UserPublicQueuesResponse>(`/users/${userId}/public-queues`)
}

// Recent Contacts
export async function getRecentContacts(limit?: number): Promise<RecentContactsListResponse> {
  const query = limit ? `?limit=${limit}` : ''
  return apiClient.get<RecentContactsListResponse>(`/users/recent-contacts${query}`)
}

File: backend/app/schemas/work_queue.py
# SPDX-FileCopyrightText: 2025 Weibo, Inc.
#
# SPDX-License-Identifier: Apache-2.0

"""Work Queue API schemas."""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field, model_validator


# Re-export enums for API use
class QueueVisibility(str, Enum):
    """Work queue visibility levels."""

    PRIVATE = "private"
    PUBLIC = "public"
    GROUP_VISIBLE = "group_visible"
    INVITE_ONLY = "invite_only"


class QueueMessageStatus(str, Enum):
    """Queue message processing status."""

    UNREAD = "unread"
    READ = "read"
    PROCESSING = "processing"
    PROCESSED = "processed"
    ARCHIVED = "archived"
    FAILED = "failed"


class QueueMessagePriority(str, Enum):
    """Queue message priority levels."""

    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"


class TriggerMode(str, Enum):
    """Auto-processing trigger mode."""

    IMMEDIATE = "immediate"
    MANUAL = "manual"
    SCHEDULED = "scheduled"
    CONDITION_BASED = "condition_based"


class ConditionAction(str, Enum):
    """Condition-based trigger action."""

    IMMEDIATE = "immediate"
    SKIP = "skip"


class ConditionType(str, Enum):
    """Condition type for auto-processing."""

    PRIORITY_HIGH = "priority_high"
    SPECIFIC_SENDER = "specific_sender"


# WorkQueue CRD schemas (stored in kinds table)
class TeamRef(BaseModel):
    """Reference to a Team for auto-processing."""

    namespace: str = "default"
    name: str


class SubscriptionRef(BaseModel):
    """Reference to a Subscription for auto-processing."""

    namespace: str = "default"
    name: str
    userId: int


class ProcessCondition(BaseModel):
    """Condition rule for auto-processing."""

    type: ConditionType
    value: Optional[str] = None  # e.g., sender user_id for SPECIFIC_SENDER
    action: ConditionAction


class AutoProcessConfig(BaseModel):
    """Auto-processing configuration."""

    enabled: bool = False
    teamRef: Optional[TeamRef] = None
    subscriptionRef: Optional[SubscriptionRef] = None
    triggerMode: TriggerMode = TriggerMode.MANUAL
    scheduleInterval: Optional[int] = Field(
        None, ge=15, description="Schedule interval in minutes (min: 15)"
    )
    conditions: Optional[List[ProcessCondition]] = None


class ResultFeedbackConfig(BaseModel):
    """Result feedback configuration."""

    replyToSender: bool = False
    saveInQueue: bool = True
    sendNotification: bool = False


class WorkQueueSpec(BaseModel):
    """WorkQueue CRD specification."""

    displayName: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    isDefault: bool = False
    visibility: QueueVisibility = QueueVisibility.PRIVATE
    visibleToGroups: Optional[List[str]] = None
    inviteCode: Optional[str] = None
    autoProcess: Optional[AutoProcessConfig] = None
    resultFeedback: Optional[ResultFeedbackConfig] = None


class WorkQueueMetadata(BaseModel):
    """WorkQueue CRD metadata."""

    name: str
    namespace: str = "default"
    displayName: Optional[str] = None


class WorkQueueStatus(BaseModel):
    """WorkQueue CRD status."""

    state: str = "Available"
    messageCount: int = 0
    unreadCount: int = 0


class WorkQueue(BaseModel):
    """WorkQueue CRD."""

    apiVersion: str = "agent.wecode.io/v1"
    kind: str = "WorkQueue"
    metadata: WorkQueueMetadata
    spec: WorkQueueSpec
    status: Optional[WorkQueueStatus] = None


class WorkQueueCreate(BaseModel):
    """Request model for creating a work queue."""

    name: str = Field(..., min_length=1, max_length=100)
    displayName: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    visibility: QueueVisibility = QueueVisibility.PRIVATE
    visibleToGroups: Optional[List[str]] = None
    autoProcess: Optional[AutoProcessConfig] = None
    resultFeedback: Optional[ResultFeedbackConfig] = None


class WorkQueueUpdate(BaseModel):
    """Request model for updating a work queue."""

    displayName: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    visibility: Optional[QueueVisibility] = None
    visibleToGroups: Optional[List[str]] = None
    autoProcess: Optional[AutoProcessConfig] = None
    resultFeedback: Optional[ResultFeedbackConfig] = None


class WorkQueueResponse(BaseModel):
    """Response model for work queue."""

    id: int
    name: str
    displayName: str
    description: Optional[str] = None
    isDefault: bool = False
    visibility: QueueVisibility
    visibleToGroups: Optional[List[str]] = None
    inviteCode: Optional[str] = None
    autoProcess: Optional[AutoProcessConfig] = None
    resultFeedback: Optional[ResultFeedbackConfig] = None
    messageCount: int = 0
    unreadCount: int = 0
    createdAt: datetime
    updatedAt: datetime


class WorkQueueListResponse(BaseModel):
    """Response model for work queue list."""

    items: List[WorkQueueResponse]
    total: int


# Queue Message schemas
class MessageContentSnapshot(BaseModel):
    """Snapshot of message content."""

    role: str  # USER or ASSISTANT
    content: str
    senderUserName: Optional[str] = None
    createdAt: Optional[str] = None
    attachments: Optional[List[Dict[str, Any]]] = None
    # IDs of subtask_contexts records pre-written from uploaded files/content.
    # Stored on the USER message that owns the attachments so that
    # _link_inbox_attachments_to_subtask() can retrieve them without a
    # separate DB column.
    attachmentContextIds: Optional[List[int]] = None


class QueueMessageCreate(BaseModel):
    """Internal model for creating a queue message."""

    queue_id: int
    sender_user_id: int
    recipient_user_id: int
    source_task_id: int
    source_subtask_ids: List[int]
    content_snapshot: List[MessageContentSnapshot]
    note: Optional[str] = None
    priority: QueueMessagePriority = QueueMessagePriority.NORMAL


class SenderInfo(BaseModel):
    """Sender information for queue message."""

    id: int
    userName: str
    email: Optional[str] = None


class QueueMessageResponse(BaseModel):
    """Response model for queue message."""

    id: int
    queueId: int
    sender: SenderInfo
    sourceTaskId: int
    sourceSubtaskIds: List[int]
    contentSnapshot: List[MessageContentSnapshot]
    note: str = ""
    priority: QueueMessagePriority
    status: QueueMessageStatus
    processResult: Dict[str, Any] = Field(default_factory=dict)
    processTaskId: int = 0
    processSubscriptionId: Optional[int] = None
    retryCount: int = 0
    createdAt: datetime
    updatedAt: datetime
    processedAt: datetime

    class Config:
        from_attributes = True


class QueueMessageListResponse(BaseModel):
    """Response model for queue message list."""

    items: List[QueueMessageResponse]
    total: int
    unreadCount: int


class QueueMessageStatusUpdate(BaseModel):
    """Request model for updating queue message status."""

    status: QueueMessageStatus


class QueueMessagePriorityUpdate(BaseModel):
    """Request model for updating queue message priority."""

    priority: QueueMessagePriority


# Batch operation schemas
class BatchMessageIds(BaseModel):
    """Request model for batch message operations."""

    messageIds: List[int] = Field(..., min_length=1, max_length=100)


class BatchStatusUpdate(BaseModel):
    """Request model for batch updating message status."""

    messageIds: List[int] = Field(..., min_length=1, max_length=100)
    status: QueueMessageStatus


class BatchOperationResult(BaseModel):
    """Response model for batch operations."""

    successCount: int
    failedCount: int
    failedIds: List[int] = Field(default_factory=list)


# Message forwarding schemas
class ForwardRecipient(BaseModel):
    """Recipient for message forwarding."""

    type: str = Field(..., description="'user' or 'group'")
    id: int = Field(..., description="User ID or Group ID")
    queueId: Optional[int] = Field(
        None, description="Queue ID (uses default if not specified)"
    )


class ForwardMessageRequest(BaseModel):
    """Request model for forwarding messages."""

    sourceTaskId: int
    subtaskIds: Optional[List[int]] = Field(
        None, description="Message IDs to forward (empty = entire conversation)"
    )
    recipients: List[ForwardRecipient]
    note: Optional[str] = Field(None, max_length=1000)
    priority: QueueMessagePriority = QueueMessagePriority.NORMAL


class ForwardMessageResponse(BaseModel):
    """Response model for message forwarding."""

    success: bool
    forwardedCount: int
    failedRecipients: Optional[List[Dict[str, Any]]] = None


# Recent contacts schemas
class RecentContactResponse(BaseModel):
    """Response model for recent contact."""

    id: int
    userId: int
    userName: str
    email: Optional[str] = None
    lastContactAt: datetime
    contactCount: int

    class Config:
        from_attributes = True


class RecentContactsListResponse(BaseModel):
    """Response model for recent contacts list."""

    items: List[RecentContactResponse]
    total: int


# User public queue schemas
class PublicQueueResponse(BaseModel):
    """Response model for public queue (for forwarding)."""

    id: int
    name: str
    displayName: str
    description: Optional[str] = None
    isDefault: bool


class UserPublicQueuesResponse(BaseModel):
    """Response model for user's public queues."""

    userId: int
    userName: str
    queues: List[PublicQueueResponse]


# Unread count schema
class UnreadCountResponse(BaseModel):
    """Response model for unread message count."""

    total: int
    byQueue: Dict[int, int] = Field(
        default_factory=dict, description="Unread count by queue ID"
    )


class ExternalSender(BaseModel):
    """External sender information for ingest API."""

    externalId: Optional[str] = None
    displayName: Optional[str] = None


class IngestSource(BaseModel):
    """Source information for ingested messages."""

    type: str = "api"
    name: Optional[str] = None


class IngestMessageRequest(BaseModel):
    """Request model for ingesting messages into a queue.

    Either `content` or `attachmentContextIds` must be provided.
    When files are uploaded via the multipart endpoint, the caller
    pre-writes them to subtask_contexts and passes the resulting IDs
    in `attachmentContextIds`, allowing `content` to be omitted.
    """

    content: Optional[str] = Field(None, max_length=50000)
    title: Optional[str] = Field(None, max_length=200)
    note: Optional[str] = Field(None, max_length=1000)
    sender: Optional[ExternalSender] = None
    attachments: Optional[List[Dict[str, Any]]] = None
    # IDs of subtask_contexts records pre-written from uploaded files
    attachmentContextIds: Optional[List[int]] = None
    source: Optional[IngestSource] = None
    priority: QueueMessagePriority = QueueMessagePriority.NORMAL

    @model_validator(mode="after")
    def validate_content_or_attachments(self) -> "IngestMessageRequest":
        """Ensure at least one of content or attachmentContextIds is provided."""
        if not self.content and not self.attachmentContextIds:
            raise ValueError(
                "Either 'content' or 'attachmentContextIds' must be provided"
            )
        return self
